core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Planner rule that matches a
+ * {@link org.apache.calcite.rel.core.Join}
+ * and expands OR clauses in join conditions.
+ *
+ * <p>For example, for the following SQL query:
+ * <pre>
+ * SELECT *
+ * FROM t1
+ * JOIN t2
+ * ON (t1.id = t2.id OR t1.name = t2.name)
+ * </pre>
+ *
+ * <p>To
+ *
+ * <pre>
+ * SELECT *
+ * FROM t1
+ * JOIN t2
+ * ON t1.id = t2.id
+ * UNION ALL
+ * SELECT *
+ * FROM t1
+ * JOIN t2
+ * ON t1.name = t2.name
+ * </pre>
+ *
+ * <p>this rule would expand the OR condition into
+ * two separate join conditions, allowing the optimizer
+ * to handle these conditions more effectively.
+ */
+@Value.Enclosing
+public class JoinConditionOrExpansionRule
+    extends RelRule<JoinConditionOrExpansionRule.Config>
+    implements TransformationRule {
+
+  /** Creates an JoinConditionExpansionOrRule. */
+  protected JoinConditionOrExpansionRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Join join = call.rel(0);
+    RelBuilder relBuilder = call.builder();
+    List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
+
+    if (orConds.size() <= 1) {
+      return;
+    }
+
+    int leftFieldCount = join.getLeft().getRowType().getFieldCount();
+    for (RexNode cond : orConds) {
+      if (!isSimpleCondition(cond, leftFieldCount)) {
```

Comment:
why is this necessary? The transformation is legal anyway.
Ideally this notion of "simple" should be parameterized, so one could plug in a different implementation.

Reply:
My initial thought is that all the expanded joins can become HashJoin. If one of them becomes NestedLoopJoin, this conversion has no advantage and CBO may not choose it. I read the jira you replied to me and it seems that there is no restriction here?

Reply:
you are thinking about this optimization only, but combined with other optimizations there could be additional benefits.
For example, this could help in combination with this new rule that is being added: https://issues.apache.org/jira/browse/CALCITE-6914

Reply:
Yes, my friend silun is working on it.